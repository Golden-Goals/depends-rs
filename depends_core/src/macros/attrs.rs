use syn::{
    parenthesized,
    parse::{Parse, ParseStream},
    parse2,
    punctuated::Punctuated,
    token::{Comma, Paren},
    Attribute,
};

/// The attribute outer token.
const DEPENDS: &str = "depends";

/// Extracts an attribute token containing multiple attribute declarations.
///
/// e.g.
///
/// #[depends(a = 1, b)]
/// #[depends(c)]
///
/// becomes
///
/// vec![a = 1, b, c]
struct DependsAttrs<T> {
    attrs: Punctuated<T, Comma>,
}

impl<T: Parse> Parse for DependsAttrs<T> {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let content;
        let _: Paren = parenthesized!(content in input);
        Ok(Self {
            attrs: content.parse_terminated(T::parse)?,
        })
    }
}

pub fn get_depends_attrs<T: Parse>(attrs: Vec<Attribute>) -> syn::Result<Vec<T>> {
    let mut res = Vec::new();
    for a in attrs {
        if a.path.is_ident(DEPENDS) {
            let dep_attrs = parse2::<DependsAttrs<T>>(a.tokens)?;
            dep_attrs.attrs.into_iter().for_each(|a| res.push(a))
        }
    }
    Ok(res)
}
