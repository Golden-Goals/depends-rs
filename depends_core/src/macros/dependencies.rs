use proc_macro2::{Ident, Span, TokenStream};
use quote::quote;
use syn::{
    parse2, spanned::Spanned, Field, Fields, GenericParam, Generics, ItemStruct, Lifetime,
    LifetimeParam, TypeParam,
};

/// In place until we figure out the dependencies generics story
fn block_generics(generics: &Generics) -> syn::Result<()> {
    if !generics.params.is_empty() {
        Err(syn::Error::new(
            generics.span(),
            "dependencies don't support generics",
        ))
    } else {
        Ok(())
    }
}

pub fn derive_dependencies(input: TokenStream) -> TokenStream {
    let ItemStruct {
        vis,
        ident,
        generics,
        fields,
        ..
    } = parse2::<ItemStruct>(input).unwrap();

    // Generics are created for derived types, so it's not clear how we
    // _could_ support them.
    block_generics(&generics).unwrap();
    let name = ident.to_string();
    let ref_ident = Ident::new(format!("{}Ref", name).as_str(), Span::call_site());
    let dep_ident = Ident::new(format!("{}Dep", name).as_str(), Span::call_site());
    let fields = if let Fields::Named(f) = fields {
        f.named
    } else {
        panic!("must be a struct with named fields");
    };
    // kind of arbitrary, but we don't have infinite alphabet and this is
    // already a large number of dependencies.
    if fields.len() < 2 {
        panic!("dependencies must have at least 2 fields");
    }
    if fields.len() > 16 {
        panic!("dependencies only supports structs with up to 16 fields");
    }

    let lifetime = LifetimeParam::new(Lifetime::new("'a", Span::call_site()));

    let mut new_fields = TokenStream::new();
    // use all fields in a dead_code block to prevent clippy from complaining
    let mut unused_fields = Vec::<TokenStream>::new();
    let mut ref_fields = TokenStream::new();
    let mut field_args = TokenStream::new();
    let mut field_new_args = TokenStream::new();
    let mut field_resolves = TokenStream::new();
    let mut dirty_field_args = Vec::<TokenStream>::new();
    let mut generics = Generics::default();
    let mut where_clauses = Vec::<TokenStream>::new();

    fields
        .into_iter()
        .zip('A'..='Z')
        .for_each(|(Field { vis, ident, ty, .. }, gen_ident)| {
            let ident = ident.expect("struct fields must be named.");
            let gen_ident = Ident::new(gen_ident.to_string().as_str(), Span::call_site());
            let new_type = quote! { ::depends::core::DepRef<#lifetime, ::std::cell::Ref<#lifetime, ::depends::core::NodeState<#ty>>> };
            generics
                .params
                .push(GenericParam::Type(TypeParam::from(gen_ident.clone())));
            new_fields.extend(quote! {
                #vis #ident: ::depends::core::Dependency<::std::rc::Rc<#gen_ident>>,
            });
            unused_fields.push(quote! {
                let _ = self.#ident;
            });
            ref_fields.extend(quote! {
                #vis #ident: #new_type,
            });
            field_args.extend(quote! {
                #ident: ::std::rc::Rc<#gen_ident>,
            });
            field_new_args.extend(quote! {
                #ident: ::depends::core::Dependency::new(#ident),
            });
            field_resolves.extend(quote! {
                #ident: self.#ident.resolve(visitor)?,
            });
            dirty_field_args.push(quote! {
                self.#ident.is_dirty()
            });
            where_clauses.push(quote! {
                for<#lifetime> #gen_ident: ::depends::core::Resolve<Output<#lifetime> = ::std::cell::Ref<#lifetime, ::depends::core::NodeState<#ty>>> + #lifetime
            });
        });

    let mut ref_generics = Generics::default();
    ref_generics
        .params
        .push(GenericParam::Lifetime(lifetime.clone()));

    quote! {
        #vis struct #dep_ident #generics {
            #new_fields
        }

        #vis struct #ref_ident #ref_generics {
            #ref_fields
        }

        impl #ident
        {
            pub fn init #generics(#field_args) -> #dep_ident #generics
            where
                #(#where_clauses),*
            {
                #dep_ident {
                    #field_new_args
                }
            }

            #[allow(dead_code)]
            fn __unused(&self) {
                #(#unused_fields);*
            }
        }

        impl #generics ::depends::core::Resolve for #dep_ident #generics
            where
                #(#where_clauses),*
        {
            type Output<#lifetime> = #ref_ident #ref_generics where Self: #lifetime;

            fn resolve(&self, visitor: &mut impl ::depends::core::Visitor) -> ::depends::core::error::ResolveResult<Self::Output<'_>> {
                Ok(#ref_ident {
                    #field_resolves
                })
            }
        }

        impl ::depends::core::IsDirty for #ref_ident <'_> {
            fn is_dirty(&self) -> bool {
                #(#dirty_field_args)||*
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use syn::parse_quote;

    use super::*;
    use crate::macros::helpers::format_source;

    #[test]
    #[ignore]
    fn test_dependencies() {
        let input = parse_quote! {
            struct Components {
                node1: Node1,
                node2: Node2,
                node3: Node3,
            }
        };

        assert_snapshot!(
            "dependencies",
            format_source(derive_dependencies(input).to_string().as_str())
        );
    }

    #[test]
    #[should_panic]
    fn test_dependencies_no_fields() {
        let input = parse_quote! {
            struct Components {
            }
        };
        derive_dependencies(input);
    }

    #[test]
    #[should_panic]
    fn test_dependencies_one_field() {
        let input = parse_quote! {
            struct Components {
                node1: Node1,
            }
        };
        derive_dependencies(input);
    }

    #[test]
    #[should_panic]
    fn test_dependencies_tuple() {
        let input = parse_quote! {
            struct Components (
                Node1,
            );
        };
        derive_dependencies(input);
    }

    #[test]
    #[should_panic]
    fn test_dependencies_enum() {
        let input = parse_quote! {
            enum Components (
                Node1 = 1;
            );
        };
        derive_dependencies(input);
    }

    #[test]
    #[should_panic]
    fn test_dependencies_too_many_fields() {
        let input = parse_quote! {
            struct Components {
                node1: Node1,
                node2: Node2,
                node3: Node3,
                node4: Node4,
                node5: Node5,
                node6: Node6,
                node7: Node7,
                node8: Node8,
                node9: Node9,
                node10: Node10,
                node11: Node11,
                node12: Node12,
                node13: Node13,
                node14: Node14,
                node15: Node15,
                node16: Node16,
                node17: Node17,
            }
        };
        derive_dependencies(input);
    }

    #[test]
    #[should_panic]
    fn test_dependencies_generics() {
        let input = parse_quote! {
            struct Components<A, B>{
                node1: A,
                node2: B,
                node3: Node3,
            }
        };
        derive_dependencies(input);
    }
}
