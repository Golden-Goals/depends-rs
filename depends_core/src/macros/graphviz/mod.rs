mod graph;
mod model;
mod parser;

pub use graph::{DerivedNodeDef, GraphvizGraph};
pub use model::{Edge, ParsedGraphvizModel};
pub use parser::GraphvizModel;
