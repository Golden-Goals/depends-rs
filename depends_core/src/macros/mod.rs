mod common;
mod dependencies;
#[cfg(test)]
pub mod helpers;
mod model;
mod value;

pub use dependencies::dependencies_attr;
pub use model::*;
pub use value::derive_value;
