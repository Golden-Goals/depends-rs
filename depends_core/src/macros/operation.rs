use proc_macro2::TokenStream;
use quote::quote;
use syn::{parse2, ItemStruct};

pub fn derive_operation(input: TokenStream) -> TokenStream {
    derive_operation_inner(input).unwrap_or_else(syn::Error::into_compile_error)
}

fn derive_operation_inner(input: TokenStream) -> syn::Result<TokenStream> {
    let ItemStruct {
        ident, generics, ..
    } = parse2::<ItemStruct>(input)?;
    let name = ident.to_string();

    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();

    Ok(quote! {
        impl #impl_generics ::depends::core::Named for #ident #ty_generics #where_clause {
            fn name() -> &'static str {
                #name
            }
        }
    })
}

#[cfg(all(test, not(miri), not(target_os = "windows")))]
mod tests {
    use insta::assert_snapshot;
    use syn::parse_quote;

    use super::*;
    use crate::macros::helpers::format_source;

    #[test]
    fn test_operation() {
        let input = parse_quote! {
            struct Operation {}
        };

        assert_snapshot!(
            "operation",
            format_source(derive_operation(input).to_string().as_str())
        );
    }

    #[test]
    fn test_operation_semi_colon() {
        let input = parse_quote! {
            struct Operation;
        };

        assert_snapshot!(
            "operation_semi_colon",
            format_source(derive_operation(input).to_string().as_str())
        );
    }

    #[test]
    fn test_operation_fields() {
        let input = parse_quote! {
            struct Operation {
                a: A,
                b: B,
            }
        };

        assert_snapshot!(
            "operation_fields",
            format_source(derive_operation(input).to_string().as_str())
        );
    }

    #[test]
    fn test_operation_tuple() {
        let input = parse_quote! {
            struct TupleOperation (
                A,
                B,
            );
        };

        assert_snapshot!(
            "operation_tuple",
            format_source(derive_operation(input).to_string().as_str())
        );
    }
}
