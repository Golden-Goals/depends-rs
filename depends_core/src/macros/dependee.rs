use proc_macro2::{Ident, Span, TokenStream};
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    parse2, Data, DeriveInput, LitBool, Token, Type,
};

use super::attrs::get_depends_attrs;

// TODO: field attr #[depends(hash_this)]

enum DependeeAttr {
    NodeName(Span, Ident),
    Dependencies(Span, Type),
    CustomClean(Span, LitBool),
}

impl Parse for DependeeAttr {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let ident = input.parse::<Ident>()?;
        input.parse::<Token![=]>()?;
        match ident.to_string().as_str() {
            "node_name" => Ok(DependeeAttr::NodeName(ident.span(), input.parse()?)),
            "dependencies" => Ok(DependeeAttr::Dependencies(ident.span(), input.parse()?)),
            "custom_clean" => Ok(DependeeAttr::CustomClean(ident.span(), input.parse()?)),
            unknown => {
                Err(syn::Error::new(
                    ident.span(),
                    format!("Unexpected attribute {:?}", unknown),
                ))
            }
        }
    }
}

struct DependeeParsedAttrs {
    node_name: Option<Ident>,
    dependencies: Option<Type>,
    custom_clean: Option<bool>,
}

impl TryFrom<Vec<DependeeAttr>> for DependeeParsedAttrs {
    type Error = syn::Error;

    fn try_from(value: Vec<DependeeAttr>) -> Result<Self, Self::Error> {
        let mut this = Self {
            node_name: None,
            dependencies: None,
            custom_clean: None,
        };
        for v in value.into_iter() {
            match v {
                DependeeAttr::Dependencies(s, deps) => {
                    if this.dependencies.is_none() {
                        this.dependencies = Some(deps);
                    } else {
                        return Err(syn::Error::new(s, "attribute specified twice"));
                    }
                }
                DependeeAttr::NodeName(s, node) => {
                    if this.node_name.is_none() {
                        this.node_name = Some(node);
                    } else {
                        return Err(syn::Error::new(s, "attribute specified twice"));
                    }
                }
                DependeeAttr::CustomClean(s, value) => {
                    if this.custom_clean.is_none() {
                        this.custom_clean = Some(value.value);
                    } else {
                        return Err(syn::Error::new(s, "attribute specified twice"));
                    }
                }
            }
        }
        Ok(this)
    }
}

pub fn derive_dependee(input: TokenStream) -> TokenStream {
    let DeriveInput {
        attrs,
        vis,
        ident,
        data,
        generics,
    } = parse2::<DeriveInput>(input).unwrap();
    let name = ident.to_string();

    // parse the attributes
    let (node_ident, dependencies_ty, custom_clean) = {
        let attrs = get_depends_attrs(attrs)
            .expect("attributes must be in the form \"#[depends(... = ..., ...)]\"");
        let parsed: DependeeParsedAttrs = attrs.try_into().unwrap();
        (
            parsed
                .node_name
                .unwrap_or_else(|| Ident::new(format!("{}Node", name).as_str(), Span::call_site())),
            parsed.dependencies.expect("missing \"dependencies\""),
            parsed.custom_clean.unwrap_or(false),
        )
    };

    let clean_clause = if custom_clean {
        TokenStream::new()
    } else {
        quote! {
            impl #generics ::depends::core::Clean for #ident {
                fn clean(&mut self) {}
            }
        }
    };

    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
    let turbo_ident = {
        let turbo = ty_generics.as_turbofish();
        quote! {
            #ident #turbo
        }
    };

    // TODO: generics aren't _actually_ supported. Dependencies must at this
    //  point in time have concrete types.
    //  It's possible in future (perhaps), that a set of dependencies can
    //  have a set of generic outputs. But importantly, these can't be tied
    //  to the type itself, as otherwise generics propagate throughout the
    //  graph implementation.
    //  So, the solution is _probably_ to specify a node by its _input_ NOT
    //  it's dependencies. Who knows. This is all very complicated.

    if let Data::Struct(_data) = data {
        quote! {
            #vis struct #node_ident #generics {
                dependencies: #dependencies_ty,
                data: ::std::cell::RefCell<::depends::core::NodeState<#ident #generics>>,
                id: usize
            }

            impl #impl_generics #node_ident #ty_generics #where_clause {
                pub fn new(dependencies: #dependencies_ty, data: #ident #ty_generics) -> ::std::rc::Rc<#node_ident #ty_generics> {
                    Self::new_with_id(dependencies, data, ::depends::core::next_node_id())
                }

                pub fn new_with_id(dependencies: #dependencies_ty, data: #ident #ty_generics, id: usize) -> ::std::rc::Rc<#node_ident #ty_generics> {
                    ::std::rc::Rc::new(
                        #node_ident {
                            dependencies,
                            data: ::std::cell::RefCell::new(::depends::core::NodeState::new(data)),
                            id
                        }
                    )
                }
            }

            impl #impl_generics ::depends::core::Identifiable for #node_ident #ty_generics #where_clause {
                fn id(&self) -> usize {
                    self.id
                }
            }

            impl #impl_generics #ident #ty_generics #where_clause {
                pub fn into_node(self, dependencies: #dependencies_ty) -> ::std::rc::Rc<#node_ident #ty_generics> {
                    #node_ident::new(
                        dependencies,
                        self,
                    )
                }
            }

            impl #impl_generics ::depends::core::Named for #ident #ty_generics #where_clause {
                fn name() -> &'static str {
                    #name
                }
            }

            impl #impl_generics ::depends::core::Named for #node_ident #ty_generics #where_clause {
                fn name() -> &'static str {
                    #turbo_ident ::name()
                }
            }

            impl #impl_generics ::depends::core::Depends for #ident #ty_generics #where_clause {
                type Input<'a>
                = <#dependencies_ty as ::depends::core::Resolve>::Output<'a>
                    where
                        Self: 'a;
            }

            impl #impl_generics ::depends::core::Resolve for #node_ident #ty_generics #where_clause {
                type Output<'a> = ::std::cell::Ref<'a, ::depends::core::NodeState<#ident>> where Self: 'a;

                fn resolve(&self, visitor: &mut impl ::depends::core::Visitor) -> Self::Output<'_> {
                    use ::depends::core::{IsDirty, Clean};

                    visitor.touch(self);
                    if visitor.visit(self) {
                        let input = self.dependencies.resolve(visitor);
                        if input.is_dirty() {
                            let mut node_state = self.data.borrow_mut();
                            node_state.clean();
                            node_state.data_mut().update_mut(input);
                            node_state.update_node_hash();
                        }
                    }
                    visitor.leave(self);
                    self.data.borrow()
                }
            }

            #clean_clause
        }
    } else {
        panic!("This macro can only be derived for structs.");
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use syn::parse_quote;

    use super::*;
    use crate::macros::helpers::format_source;

    #[test]
    fn test_dependee() {
        let input: TokenStream = parse_quote! {
            #[depends(node_name = SomeName, dependencies = SomeType)]
            #[diesel(some_arg)]
            pub struct Foo {
                #[depends(hi)]
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "dependee",
            format_source(derive_dependee(input).to_string().as_str())
        );
    }

    #[test]
    fn test_dependee_generics() {
        let input = parse_quote! {
            #[depends(dependencies = SomeDeps<D, E>)]
            struct Foo<T> {
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "dependee_generics",
            format_source(derive_dependee(input).to_string().as_str())
        );
    }

    #[test]
    fn test_dependee_generics_custom_clean() {
        let input = parse_quote! {
            #[depends(dependencies = SomeDeps<D, E>, custom_clean = true)]
            struct Foo<T> {
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "dependee_generics_custom_clean",
            format_source(derive_dependee(input).to_string().as_str())
        );
    }

    #[test]
    fn test_dependee_single_dependency() {
        let input = parse_quote! {
            #[depends(dependencies = Dependency<Rc<SomeNode<Bar>>>, custom_clean = true)]
            struct Foo<T> {
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "dependee_single_dependency",
            format_source(derive_dependee(input).to_string().as_str())
        );
    }
}
