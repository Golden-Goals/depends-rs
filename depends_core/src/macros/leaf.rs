use proc_macro2::{Ident, Span, TokenStream};
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    parse2, Data, DeriveInput, LitBool, Token,
};

use super::attrs::get_depends_attrs;

enum LeafAttr {
    CanHash(Span, LitBool),
    CustomClean(Span, LitBool),
}

impl Parse for LeafAttr {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let ident = input.parse::<Ident>()?;
        input.parse::<Token![=]>()?;
        match ident.to_string().as_str() {
            "can_hash" => Ok(LeafAttr::CanHash(ident.span(), input.parse()?)),
            "custom_clean" => Ok(LeafAttr::CustomClean(ident.span(), input.parse()?)),
            unknown => {
                Err(syn::Error::new(
                    ident.span(),
                    format!("Unexpected attribute {:?}", unknown),
                ))
            }
        }
    }
}

struct LeafParsedAttrs {
    can_hash: Option<bool>,
    custom_clean: Option<bool>,
}

impl TryFrom<Vec<LeafAttr>> for LeafParsedAttrs {
    type Error = syn::Error;

    fn try_from(value: Vec<LeafAttr>) -> Result<Self, Self::Error> {
        let mut this = Self {
            can_hash: None,
            custom_clean: None,
        };
        for v in value.into_iter() {
            match v {
                LeafAttr::CanHash(s, value) => {
                    if this.can_hash.is_none() {
                        this.can_hash = Some(value.value);
                    } else {
                        return Err(syn::Error::new(s, "attribute specified twice"));
                    }
                }
                LeafAttr::CustomClean(s, value) => {
                    if this.custom_clean.is_none() {
                        this.custom_clean = Some(value.value);
                    } else {
                        return Err(syn::Error::new(s, "attribute specified twice"));
                    }
                }
            }
        }
        Ok(this)
    }
}

pub fn derive_leaf(input: TokenStream) -> TokenStream {
    let DeriveInput {
        ident,
        data,
        generics,
        attrs,
        ..
    } = parse2::<DeriveInput>(input).unwrap();
    let name = ident.to_string();
    let (can_hash, custom_clean) = {
        let attrs = get_depends_attrs(attrs)
            .expect("attributes must be in the form \"#[depends(... = ..., ...)]\"");
        let parsed: LeafParsedAttrs = attrs.try_into().unwrap();
        (
            parsed.can_hash.unwrap_or(true),
            parsed.custom_clean.unwrap_or(false),
        )
    };

    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();

    let clean_clause = if custom_clean {
        TokenStream::new()
    } else {
        quote! {
            impl #ty_generics ::depends::core::Clean for #ident #ty_generics #where_clause {
                fn clean(&mut self) {}
            }
        }
    };

    let hash_value_clause = if can_hash {
        quote! {
            fn hash_value(&self, hasher: &mut impl ::std::hash::Hasher) -> ::depends::core::NodeHash {
                ::depends::core::NodeHash::Hashed({
                    self.hash(hasher);
                    hasher.finish()
                })
            }
        }
    } else {
        quote! {
            fn hash_value(&self, _: &mut impl ::std::hash::Hasher) -> ::depends::core::NodeHash {
                ::depends::core::NodeHash::NotHashed
            }
        }
    };

    if let Data::Struct(_) = data {
        quote! {
            impl #impl_generics ::depends::core::Named for #ident #ty_generics #where_clause {
                fn name() -> &'static str {
                    #name
                }
            }

            impl #impl_generics #ident #ty_generics #where_clause {
                pub fn into_leaf(self) -> ::std::rc::Rc<::depends::core::LeafNode<Self>> {
                    ::depends::core::LeafNode::new(self)
                }
            }

            impl #impl_generics ::depends::core::HashValue for #ident #ty_generics #where_clause {
                #hash_value_clause
            }

            #clean_clause
        }
    } else {
        panic!("This macro can only be derived for structs.");
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use syn::parse_quote;

    use super::*;
    use crate::macros::helpers::format_source;

    #[test]
    fn test_leaf() {
        let input: TokenStream = parse_quote! {
            #[diesel(some_arg)]
            pub struct Foo {
                #[depends(hi)]
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "leaf",
            format_source(derive_leaf(input).to_string().as_str())
        );
    }

    #[test]
    fn test_leaf_generics() {
        let input = parse_quote! {
            struct Foo<T> {
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "leaf_generics",
            format_source(derive_leaf(input).to_string().as_str())
        );
    }

    #[test]
    fn test_leaf_generics_custom_clean() {
        let input = parse_quote! {
            #[depends(custom_clean = true)]
            struct Foo<T> {
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "leaf_generics_custom_clean",
            format_source(derive_leaf(input).to_string().as_str())
        );
    }

    #[test]
    #[should_panic]
    fn test_leaf_on_enum() {
        let input = parse_quote! {
            enum Foo {
                Thing(usize)
            }
        };

        derive_leaf(input);
    }
}
