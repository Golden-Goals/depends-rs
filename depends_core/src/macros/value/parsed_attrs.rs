use super::{field_attrs::ValueFieldAttr, struct_attrs::ValueStructAttr, ValueAttrModel};
use crate::macros::{common::duplicate_attribute, HashLogic};

pub struct ValueParsedAttrs {
    pub hashing: Option<HashLogic>,
    pub custom_clean: Option<bool>,
}

impl TryFrom<ValueAttrModel> for ValueParsedAttrs {
    type Error = syn::Error;

    fn try_from(attrs: ValueAttrModel) -> Result<Self, Self::Error> {
        let mut this = Self {
            custom_clean: None,
            hashing: None,
        };
        for v in attrs.struct_attrs.into_iter() {
            match v {
                ValueStructAttr::Unhashable(s) => {
                    if this.hashing.is_none() {
                        this.hashing = Some(HashLogic::Unhashable);
                    } else {
                        return Err(duplicate_attribute(s));
                    }
                }
                ValueStructAttr::CustomClean(s) => {
                    if this.custom_clean.is_none() {
                        this.custom_clean = Some(true);
                    } else {
                        return Err(duplicate_attribute(s));
                    }
                }
            }
        }
        for v in attrs.field_attrs.into_iter() {
            for a in v.field_attrs {
                let ValueFieldAttr::Hash(s) = a;
                if this.hashing.is_none() {
                    this.hashing = Some(HashLogic::Field(v.ident.clone()));
                } else {
                    return Err(duplicate_attribute(s));
                }
            }
        }
        Ok(this)
    }
}
