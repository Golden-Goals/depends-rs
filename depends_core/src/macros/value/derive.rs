use proc_macro2::TokenStream;
use quote::quote;
use syn::{parse2, spanned::Spanned, Field, ItemStruct};

use super::parsed_attrs::ValueParsedAttrs;
use crate::macros::{
    model::{get_depends_attrs, FieldAttrs},
    value::ValueAttrModel,
    HashLogic,
};

pub fn derive_value(input: TokenStream) -> TokenStream {
    derive_value_inner(input).unwrap_or_else(syn::Error::into_compile_error)
}

fn derive_value_inner(input: TokenStream) -> syn::Result<TokenStream> {
    let ItemStruct {
        ident,
        fields,
        generics,
        attrs,
        ..
    } = parse2::<ItemStruct>(input)?;

    let name = ident.to_string();
    let (custom_clean, hashing) = {
        let struct_attrs = get_depends_attrs(&attrs)?;
        let field_attrs: syn::Result<Vec<_>> = fields
            .iter()
            .map(|Field { attrs, ident, .. }| {
                Ok(FieldAttrs {
                    ident: ident
                        .as_ref()
                        .ok_or_else(|| {
                            syn::Error::new(
                                ident.span(),
                                "Only structs with named fields are supported.",
                            )
                        })?
                        .clone(),
                    field_attrs: get_depends_attrs(attrs)?,
                })
            })
            .collect();
        let field_attrs = field_attrs?;

        let model = ValueAttrModel {
            struct_attrs,
            field_attrs,
        };
        let parsed: ValueParsedAttrs = model.try_into()?;
        (
            parsed.custom_clean.unwrap_or(false),
            parsed.hashing.unwrap_or(HashLogic::Struct),
        )
    };

    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();

    let clean_clause = if custom_clean {
        TokenStream::new()
    } else {
        quote! {
            impl #ty_generics ::depends::core::Clean for #ident #ty_generics #where_clause {
                fn clean(&mut self) {}
            }
        }
    };

    let hash_value_clause = hashing.to_tokens();

    Ok(quote! {
        impl #impl_generics ::depends::core::Named for #ident #ty_generics #where_clause {
            fn name() -> &'static str {
                #name
            }
        }

        impl #impl_generics ::depends::core::HashValue for #ident #ty_generics #where_clause {
            fn hash_value(&self, hasher: &mut impl ::std::hash::Hasher) -> ::depends::core::NodeHash {
                use ::std::hash::Hash;
                #hash_value_clause
            }
        }

        #clean_clause
    })
}

#[cfg(all(test, not(miri), not(target_os = "windows")))]
mod tests {
    use insta::assert_snapshot;
    use syn::parse_quote;

    use super::*;
    use crate::macros::helpers::format_source;

    #[test]
    fn test_input() {
        let input: TokenStream = parse_quote! {
            #[diesel(some_arg)]
            pub struct Foo {
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "value",
            format_source(derive_value(input).to_string().as_str())
        );
    }

    #[test]
    fn test_input_hashed() {
        let input = parse_quote! {
            struct Foo<T> {
                bar: Vec<usize>,
                #[depends(hash)]
                number: usize,
            }
        };
        assert_snapshot!(
            "value_hashed_attr",
            format_source(derive_value(input).to_string().as_str())
        );
    }

    #[test]
    fn test_input_unhashable() {
        let input = parse_quote! {
            #[depends(unhashable)]
            struct Foo<T> {
                bar: Vec<usize>,
                number: usize,
            }
        };
        assert_snapshot!(
            "value_unhashable",
            format_source(derive_value(input).to_string().as_str())
        );
    }

    #[test]
    fn test_input_generics_custom_clean() {
        let input = parse_quote! {
            #[depends(custom_clean)]
            struct Foo<T> {
                bar: Vec<usize>
            }
        };
        assert_snapshot!(
            "value_generics_custom_clean",
            format_source(derive_value(input).to_string().as_str())
        );
    }
}
