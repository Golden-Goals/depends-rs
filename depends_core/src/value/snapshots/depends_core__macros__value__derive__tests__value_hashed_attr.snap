---
source: depends_core/src/macros/value/derive.rs
assertion_line: 134
expression: format_source(derive_value(input).to_string().as_str())
---
impl<T> ::depends::Named for Foo<T> {
    fn name() -> &'static str {
        "Foo"
    }
}
impl<T> ::depends::HashValue for Foo<T> {
    fn hash_value(&self, hasher: &mut impl ::std::hash::Hasher) -> ::depends::NodeHash {
        use ::std::hash::Hash;
        ::depends::NodeHash::Hashed({
            self.number.hash(hasher);
            hasher.finish()
        })
    }
}
impl<T> ::depends::Clean for Foo<T> {
    fn clean(&mut self) {}
}

